crates/driver/src/logic/competition/solution/mod.rs

```diff
@@ -1,40 +1,145 @@
 use {
-    super::auction::Auction,
     crate::{
+        logic::{competition, eth},
         solver::{self, Solver},
+        util,
         Ethereum,
         Simulator,
     },
-    primitive_types::{H160, U256},
+    num::ToPrimitive,
+    primitive_types::U256,
     std::collections::HashMap,
 };
 
-mod allowances;
+mod approvals;
 pub mod settlement;
 
-pub use {
-    allowances::Allowances,
-    settlement::{Score, Settlement},
-};
+pub use {approvals::Approvals, settlement::Settlement};
 
 /// A solution represents a set of orders which the solver has found an optimal
 /// way to settle. A [`Solution`] is generated by a solver as a response to a
 /// [`super::auction::Auction`].
 #[derive(Debug)]
 pub struct Solution {
-    pub prices: HashMap<H160, U256>,
-    pub allowances: Allowances,
+    // TODO After #831 we hope to make this a vec and not a hashmap
+    pub orders: HashMap<usize, Order>,
+    pub jit_orders: Vec<JitOrder>,
+    // TODO This will be changed in #831:
+    // 1. It should not be a HashMap, it should be a list.
+    // 2. It should be part of the orders field, as an enum
+    // But I'll make it work like this for now
+    pub amms: HashMap<eth::Address, Vec<Amm>>,
+    pub prices: HashMap<eth::Token, eth::Ether>,
+    pub approvals: Approvals,
+    pub interactions: Vec<SafeInteraction>,
+}
+
+/// An interaction with inputs and outputs which can be used to ensure that the
+/// solver isn't cheating.
+#[derive(Debug)]
+pub struct SafeInteraction {
+    pub inner: eth::Interaction,
+    pub inputs: Vec<eth::Asset>,
+    pub outputs: Vec<eth::Asset>,
+}
+
+#[derive(Debug)]
+pub struct Amm {
+    pub sell: eth::Asset,
+    pub buy: eth::Asset,
+    pub sequence: u32,
+    pub position: u32,
+    pub internal: bool,
+}
+
+// TODO Why is this on JIT orders in the SOLUTION and also on all orders in the
+// auction?
+#[derive(Debug)]
+pub enum SellSource {
+    /// Direct ERC20 allowances to the Vault relayer contract
+    Erc20,
+    /// ERC20 allowances to the Vault with GPv2 relayer approval
+    Internal,
+    /// Internal balances to the Vault with GPv2 relayer approval
+    External,
+}
+
+#[derive(Debug)]
+pub enum BuyDestination {
+    /// Direct ERC20 allowances to the Vault relayer contract
+    Erc20,
+    /// ERC20 allowances to the Vault with GPv2 relayer approval
+    Internal,
+}
+
+/// An order which was executed as part of this solution.
+#[derive(Debug)]
+pub struct Order {
```

Comment:
I think this can be simplified a bit. The only information that we need from a solver is "how much was executed". Everything else is an inherent property to the order that doesn't need to be echoed here.